File: pyorchestrator/backend/app/schemas/updates.py
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class UpdateStepResponse(BaseModel):
    id: str
    label: str
    status: str = "pending"
    message: str | None = None
    at: datetime | None = None

    model_config = {"from_attributes": True}


class UpdateJobResponse(BaseModel):
    id: UUID
    status: str
    from_version: str
    target_version: str
    target_tag: str
    release_url: str | None = None
    release_notes: str | None = None
    steps: list[UpdateStepResponse] = Field(default_factory=list)
    rollback_snapshot: dict | None = None
    trigger: str
    triggered_by_user_id: UUID | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    error: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class UpdateSettingsResponse(BaseModel):
    check_enabled: bool
    notify_enabled: bool
    auto_update_enabled: bool
    check_interval_hours: int
    auto_update_interval_hours: int
    github_repo: str
    include_prerelease: bool
    last_check_at: str | None = None
    last_known_latest_version: str | None = None
    last_notified_version: str | None = None
    dismissed_version: str | None = None
    last_applied_version: str | None = None


class UpdateSettingsUpdateRequest(BaseModel):
    check_enabled: bool | None = None
    notify_enabled: bool | None = None
    auto_update_enabled: bool | None = None
    check_interval_hours: int | None = Field(default=None, ge=1, le=168)
    auto_update_interval_hours: int | None = Field(default=None, ge=1, le=720)
    github_repo: str | None = None
    include_prerelease: bool | None = None


class UpdateCheckResponse(BaseModel):
    current_version: str
    latest_version: str | None = None
    latest_tag: str | None = None
    update_available: bool
    release_url: str | None = None
    release_notes: str | None = None
    published_at: str | None = None
    checked_at: str
    executor_available: bool
    executor_reason: str | None = None
    deploy_mode: str


class UpdateStatusResponse(UpdateCheckResponse):
    settings: UpdateSettingsResponse
    active_job: UpdateJobResponse | None = None
    recent_jobs: list[UpdateJobResponse] = Field(default_factory=list)
    show_notification: bool = False


class UpdateApplyRequest(BaseModel):
    target_version: str | None = None
    target_tag: str | None = None


class UpdateDismissRequest(BaseModel):
    version: str


class UpdateCheckResponseLegacy(BaseModel):
    current_version: str
    latest_version: str | None
    update_available: bool
    release_notes: str = ""

File: dashboard/src/api/updates.ts
import { fetchWithAuth } from './client';
import { tGlobal } from '../i18n/runtime';

export type UpdateComponentId = 'crm' | 'dynamic-api' | 'pyorchestrator';

export type StepStatus = 'pending' | 'running' | 'completed' | 'failed' | 'skipped';

export type JobStatus = 'queued' | 'running' | 'completed' | 'failed';

export interface UpdateStep {
  id: string;
  label: string;
  status: StepStatus;
  message?: string;
  startedAt?: string;
  finishedAt?: string;
}

export interface UpdateJob {
  id: string;
  component: UpdateComponentId;
  targetVersion: string;
  fromVersion: string;
  status: JobStatus;
  steps: UpdateStep[];
  logs: string[];
  error?: string;
  createdAt: string;
  finishedAt?: string;
}

export interface ComponentCheck {
  id: UpdateComponentId;
  label: string;
  githubRepo: string;
  currentVersion: string;
  latestVersion: string | null;
  latestTag: string | null;
  updateAvailable: boolean;
  releaseUrl: string | null;
  releaseNotes: string | null;
  publishedAt: string | null;
  checkedAt: string | null;
}

export interface UpdatesStatus {
  executorAvailable: boolean;
  executorReason: string | null;
  lastCheckAt: string | null;
  components: ComponentCheck[];
  activeJob: UpdateJob | null;
  recentJobs: UpdateJob[];
  showNotification: boolean;
  dismissedFailedJobIds: Record<string, string>;
}

async function parseJson<T>(res: Response): Promise<T> {
  const json = (await res.json()) as { success?: boolean; error?: string; data?: T };
  if (!res.ok || json.success === false) {
    throw new Error(json.error || res.statusText || tGlobal('errors.requestFailed', { status: res.status }));
  }
  return json.data as T;
}

export async function getUpdatesStatus(refresh = false): Promise<UpdatesStatus> {
  const q = refresh ? '?refresh=1' : '';
  const res = await fetchWithAuth(`/api/crm/updates/status${q}`);
  return parseJson(res);
}

export async function checkForUpdates(): Promise<ComponentCheck[]> {
  const res = await fetchWithAuth('/api/crm/updates/check', { method: 'POST' });
  const data = await parseJson<{ components: ComponentCheck[] }>(res);
  return data.components;
}

export async function applyUpdate(component: UpdateComponentId, targetTag?: string): Promise<UpdateJob> {
  const res = await fetchWithAuth(`/api/crm/updates/apply/${component}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ targetTag }),
  });
  return parseJson(res);
}

export async function dismissUpdate(
  component: UpdateComponentId,
  version: string,
  jobId?: string
): Promise<void> {
  const res = await fetchWithAuth(`/api/crm/updates/dismiss/${component}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ version, jobId }),
  });
  await parseJson(res);
}

export function componentById(status: UpdatesStatus | null, id: UpdateComponentId): ComponentCheck | undefined {
  return status?.components.find((c) => c.id === id);
}

/** Активная задача, блокирующая баннер и быстрый опрос (игнорирует зависшие queued после успешного обновления). */
export function isBlockingActiveJob(status: UpdatesStatus | null | undefined): boolean {
  const job = status?.activeJob;
  if (!job) return false;

  const comp = status.components.find((c) => c.id === job.component);
  const targetAlreadyInstalled = comp?.currentVersion === job.targetVersion;
  const neverStarted =
    job.status === 'queued' && job.steps.every((step) => step.status === 'pending');

  if (targetAlreadyInstalled && neverStarted) return false;

  return job.status === 'queued' || job.status === 'running';
}

export type RepairIssueSeverity = 'ok' | 'warning' | 'error';

export interface RepairIssue {
  code: string;
  severity: RepairIssueSeverity;
  detail?: string;
  fixId?: string;
}

export interface RepairPaths {
  deployRoot: string;
  hostProjectRoot: string;
  detectedHostRoot: string | null;
  hostDataDir: string;
  envFile: string;
  envWashHostRoot: string | null;
  envDataDir: string | null;
}

export interface RepairDiagnoseResult {
  checkedAt: string;
  paths: RepairPaths;
  issues: RepairIssue[];
  healthy: boolean;
}

export interface RepairApplyResult {
  applied: string[];
  failed: Array<{ action: string; error: string }>;
  logs: string[];
  diagnose: RepairDiagnoseResult;
}

export async function diagnoseIntegrity(): Promise<RepairDiagnoseResult> {
  const res = await fetchWithAuth('/api/crm/updates/repair');
  return parseJson(res);
}

export async function applyIntegrityRepair(actions: string[]): Promise<RepairApplyResult> {
  const res = await fetchWithAuth('/api/crm/updates/repair', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ actions }),
  });
  return parseJson(res);
}
